sysboundary/pml.cpp

```diff
@@ -0,0 +1,275 @@
+#include <iostream>
+#include <iomanip> 
+#include <cmath>
+#include <sstream>
+#include <ctime>
+#include <omp.h>
+#include "fsgrid.hpp"
+#include "../definitions.h"
+#include "../common.h"
+#include "../parameters.h"
+#include "../readparameters.h"
+
+
+
+
+bool buildPMLGrids(
+    FsGrid<std::array<Real, fsgrids::pml::N_PML>, 2> &pmlGrid)
+{
+
+   typedef Parameters P;
+
+
+   /*----Get PML Array and Domain Size-----*/
+   std::array<Real, fsgrids::pml::N_PML> *pmlValue;
+   int start = 2;   /*don't judge will change it */
+   std::array<int32_t, 3> pos;
+   const int *pmlDims = &pmlGrid.getLocalSize()[0];
+   const int *globalDims = &pmlGrid.getGlobalSize()[0];
+
+
+
+   /*Read PLM faces to build*/
+   std::vector<std::string> thisSpeciesFaceList;
+   Readparameters::get("PML.face", thisSpeciesFaceList) ;
+   bool Xp, Xm, Yp, Ym, Zp, Zm = false;
+   for (auto &face : thisSpeciesFaceList){
+   if(face == "x+") {Xp = true;}      //set face to true  
+   if(face == "x-") {Xm = true;}      //set face to true
+   if(face == "y+") {Yp = true;}      //set face to true
+   if(face == "y-") {Ym = true;}      //set face to true
+   if(face == "z+") {Zp = true;}      //set face to true
+   if(face == "z-") {Zm = true;}      //set face to true  
+   }   
+
+
+    /*-----Initially set all arrays to one-----*/
+    /*Iterate over domain and set the PML arrays to 1.0 thus not affecting the fieldsolver*/ 
+    #pragma omp parallel for collapse(3)
+    for (int kk = 0; kk < pmlDims[2]; kk++){
+        for (int jj = 0; jj < pmlDims[1]; jj++){
+            for (int ii = 0; ii < pmlDims[0]; ii++){
+
+                pmlValue = pmlGrid.get(ii, jj, kk);
+                pmlValue->at(fsgrids::pml::PGI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFI3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFJ3) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PGK3) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK1) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK2) = 1.0;
+                pmlValue->at(fsgrids::pml::PFK3) = 1.0;
+            }
+        }
+    }
+
+
+   Real xnum, xd;
+   Real xxn, xn;
+
+   // Attentuation Along the X-Dimension
+   #pragma omp parallel for collapse(3)
+   for (int kk = 0; kk < pmlDims[2] ; kk++){
+      for (int jj = 0; jj < pmlDims[1]; jj++){
+         for (int ii = 0; ii <pmlDims[0]; ii++){
+
+         // Get Global Arrays
+         pos=pmlGrid.getGlobalIndices(ii,jj,kk);
+
+         if (Xm==true){
+            if (pos[0]>=start && pos[0]<=P::pmlWidthX+start){
```

Comment:
These two ifs can be combined but the compiler will probably do it anyway.

This `pos` also needs to be thread-private. Do check all the variables. Or remove threading for now and we can do it at a later stage.